SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialLoadBalanceEnforcer.cpp

```diff
@@ -1,271 +1,215 @@
 		{
-			UE_LOG(LogSpatialLoadBalanceEnforcer, Log,
-				   TEXT("Failed to update the EntityACL to match the authority intent; this worker lost authority over the EntityACL since "
-						"the request was queued."
-						" Source worker ID: %s. Entity ID %lld. Desination worker ID: %s."),
-				   *WorkerId, EntityId, **DestinationWorkerId);
-			CompletedRequests.Add(EntityId);
+			Acl.ComponentWriteAcl.Add(ComponentId, { SpatialConstants::UnrealServerAttributeSet });
 			continue;
 		}
 
-		TArray<Worker_ComponentId> ComponentIds;
-
-		EntityAcl* Acl = StaticComponentView->GetComponentData<EntityAcl>(EntityId);
-		Acl->ComponentWriteAcl.GetKeys(ComponentIds);
-
-		// Ensure that every component ID in ComponentPresence is set in the write ACL.
-		for (const auto& RequiredComponentId : ComponentPresenceComponent->ComponentList)
-		{
-			ComponentIds.AddUnique(RequiredComponentId);
-		}
-
-		// Get the client worker ID net-owning this Actor from the NetOwningClientWorker.
-		PhysicalWorkerName PossessingClientId =
-			NetOwningClientWorkerComponent->WorkerId.IsSet() ? NetOwningClientWorkerComponent->WorkerId.GetValue() : FString();
-
-		PendingRequests.Push(
-			AclWriteAuthorityRequest{ EntityId, *DestinationWorkerId, Acl->ReadAcl, { { PossessingClientId } }, ComponentIds });
-
-		CompletedRequests.Add(EntityId);
+		Acl.ComponentWriteAcl.Add(ComponentId, { OwningServerWorkerAttributeSet });
 	}
 
-	AclWriteAuthAssignmentRequests.RemoveAll([CompletedRequests](const Worker_EntityId& EntityId) {
-		return CompletedRequests.Contains(EntityId);
-	});
-
-	return PendingRequests;
+	const FWorkerComponentUpdate Update = Acl.CreateEntityAclUpdate();
+	return EntityComponentUpdate{ EntityId, ComponentUpdate(OwningComponentUpdatePtr(Update.schema_type), Update.component_id) };
 }
 
-void SpatialLoadBalanceEnforcer::QueueAclAssignmentRequest(const Worker_EntityId EntityId)
+void SpatialLoadBalanceEnforcer::PopulateDataStore(const Worker_EntityId EntityId)
 {
-	UE_LOG(LogSpatialLoadBalanceEnforcer, Verbose, TEXT("Queueing ACL assignment request for entity %lld on worker %s."), EntityId,
-		   *WorkerId);
-	AclWriteAuthAssignmentRequests.Add(EntityId);
+	LBComponents& Components = DataStore.Emplace(EntityId, LBComponents{});
+	for (auto& ComponentData : SubView->GetView()[EntityId].Components)
+	{
+		switch (ComponentData.GetComponentId())
+		{
+		case SpatialConstants::ENTITY_ACL_COMPONENT_ID:
+			Components.Acl = EntityAcl(ComponentData.GetUnderlying());
+			break;
+		case SpatialConstants::AUTHORITY_INTENT_COMPONENT_ID:
+			Components.Intent = AuthorityIntent(ComponentData.GetUnderlying());
+			break;
+		case SpatialConstants::COMPONENT_PRESENCE_COMPONENT_ID:
+			Components.Presence = ComponentPresence(ComponentData.GetUnderlying());
+			break;
+		case SpatialConstants::NET_OWNING_CLIENT_WORKER_COMPONENT_ID:
+			Components.OwningClientWorker = NetOwningClientWorker(ComponentData.GetUnderlying());
+			break;
+		default:
```

Comment:
Do we expect this case to be hit? Will the subview contain any other components?

Reply:
we do expect this case to be hit, as we are grabbing the state of the entity form the main view, so contains all components. The components are stored as an array in the main view, so it is quicker to loop through like this once than find each individually.

Reply:
Sweet, if it's an array that's coool.